src/tape/ensemble.py

```diff
@@ -453,21 +453,30 @@ def coalesce(self, input_cols, output_col, table="object", drop_inputs=False):
         else:
             raise ValueError(f"{table} is not one of 'object' or 'source'")
 
-        # Create a subset dataframe with the coalesced columns
-        # Drop index for dask series operations - unfortunate
-        coal_ddf = table_ddf[input_cols].reset_index()
-
-        # Coalesce each column iteratively
-        i = 0
-        coalesce_col = coal_ddf[input_cols[0]]
-        while i < len(input_cols) - 1:
-            coalesce_col = coalesce_col.combine_first(coal_ddf[input_cols[i + 1]])
-            i += 1
-        # Assign the new column to the subset df, and reintroduce index
-        coal_ddf = coal_ddf.assign(**{output_col: coalesce_col}).set_index(self._id_col)
-
-        # assign the result to the desired column name
-        table_ddf = table_ddf.assign(**{output_col: coal_ddf[output_col]})
+        def coalesce_partition(df, input_cols, output_col):
+            """Coalescing function for a single partition (pandas dataframe)"""
+
+            # Create a subset dataframe per input column
+            # Rename column to output to allow combination
+            input_dfs = []
+            for col in input_cols:
+                col_df = df[[col]]
+
+                input_dfs.append(col_df.rename(columns={col: output_col}))
+
+            # Combine each dataframe
+            i = 0
+            coal_df = input_dfs[0]
+            while i < len(input_dfs) - 1:
+                coal_df = coal_df.combine_first(input_dfs[i + 1])
+                i += 1
```

Comment:
Nit: not relevant for this specific PR, but we could alternatively

            # Combine each dataframe
            coal_df = input_dfs.pop()
            while input_dfs:
                coal_df = coal_df.combine_first(input_dfs.pop())

Using `pop(0)` if we care about preserving the current order

This seems a bit more readable to me but up to you

Reply:
Nice, this looks better to me as well, implemented!